clang/test/SemaSYCL/unique_stable_name.cpp

```diff
@@ -0,0 +1,174 @@
+// RUN: %clang_cc1 %s --std=c++17 -fsyntax-only -fsycl-is-device -Wno-unused -verify
+
+template <typename KernelName, typename KernelType>
+[[clang::sycl_kernel]] void kernel_single_task(KernelType kernelFunc) { // #1 #2 #3 #4 #5 #6 #7 #8 #9 #10 #11
+  kernelFunc();
+}
+
+template <typename Func>
+void kernel3func(const Func &F3) {
+  constexpr const char *F3_output = __builtin_unique_stable_name(F3);
+  // expected-error@#8{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
+  // expected-note@+1{{in instantiation of function template specialization}}
+  kernel_single_task<class kernel3>(F3); // 8th instantiation throwing error
+}
+
+template <typename Func>
+void kernel4func(const Func &F4) {
+  constexpr const char *F4_output = __builtin_unique_stable_name(F4);
+  // expected-error@#9{{kernel instantiation changes the result of an evaluated '__builtin_unique_stable_name'}}
```

Comment:
@erichkeane -- this one surprises me as well. We compute the stable name for `F4` but then get an error when passing an empty lambda to the kernel.

I'm starting to wonder if the same lambda signature is being computed as the same builtin stable name, despite being unique lambda instances. e.g., passing an empty lambda to a kernel means you can never pass another empty lambda to a kernel again within the same TU because that'll think it's computing the same unique stable name twice.

Reply:
The issue in this one is that the empty lambda is named based on its containing function, which is kernel4func.  F4 is a PART of that name, so its unique-stable-name mangling changes there.

Reply:
Oh yeah, this same case has tripped me up twice now. @schittir -- please be sure to add comments to this test case explaining what's going on, otherwise I'm sure my forgetful self will ask again in a week. :-D